### PR TITLE
Improve Redshift

### LINKED ISSUE #14697: Redshift connector deletes do not delete data
The documentation states that delete operations for Redshift connector work only if the predicate in the clause can be fully pushed down to the data source. However, in my case I can see the same delete query with predicates is being sent to Redshift and the data is not deleted. 

Trino-cli
Screenshot 2022-09-27 at 4 05 33 PM
Redshift console 
Screenshot 2022-09-27 at 4 07 35 PM

https://trinodb.slack.com/archives/CGB0QHWSW/p1664275183300789

### PR DESCRIPTION
**Pull request by dain**
## Description
This PR contains many improvements for Redshift with a full test suite.  Currently the test suite must be run manually.

I have run the test manually and they all pass.

Fixes #14697

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Redshift
* Add aggregation, join, and top N pushdown ({issue}`15365`)
* Implement DELETE ({issue}`15365`)
* Implement full type mapping ({issue}`15365`)
* Add schema, table and column name length checks ({issue}`15365`)
```

### CODE REVIEW

File: plugin/trino-redshift/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+To run the Redshift tests you will need to provision a Redshift cluster.  The
+tests are designed to run on the smallest possible Redshift cluster containing
+is a single dc2.large instance. Additionally, you will need a S3 bucket 
+containing TPCH tiny data in Parquet format.  The files should be named:
+
+```
+s3://<your_bucket>/tpch/tiny/<table_name>.parquet
+```
+
+To run the sets set the following system properties:
```
Comment (electrum) [inline plugin/trino-redshift/README.md:12]: sets -> tests

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClientModule.java [hunk 2]
```diff
@@ -45,6 +47,14 @@ public void configure(Binder binder)
     @ForBaseJdbc
     public static ConnectionFactory getConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
     {
-        return new DriverConnectionFactory(new Driver(), config, credentialProvider);
+        return new DriverConnectionFactory(new Driver(), config.getConnectionUrl(), getDriverProperties(), credentialProvider);
+    }
+
+    private static Properties getDriverProperties()
```
Comment (electrum) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClientModule.java:61]: Is this related to adding tests?

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClient.java [hunk 3]
```diff
@@ -130,6 +131,42 @@ public PreparedStatement getPreparedStatement(Connection connection, String sql)
         return statement;
     }
 
+    @Override
+    protected void verifySchemaName(DatabaseMetaData databaseMetadata, String schemaName)
+            throws SQLException
+    {
+        // Redshift truncates schema name to 127 chars silently
+        if (schemaName.length() > databaseMetadata.getMaxSchemaNameLength()) {
+            throw new TrinoException(
```
Comment (electrum) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClient.java:443]: Nit:  I wouldn't wrap here

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClient.java [hunk 4]
```diff
@@ -130,6 +131,42 @@ public PreparedStatement getPreparedStatement(Connection connection, String sql)
...
+                    NOT_SUPPORTED,
+                    format("Schema name must be shorter than or equal to '%s' characters but got '%s'", databaseMetadata.getMaxSchemaNameLength(), schemaName.length()));
```
Comment (electrum) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClient.java:445]: This could use the `formatted` method

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftTableStatisticsReader.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class RedshiftTableStatisticsReader
+{
+    private final ConnectionFactory connectionFactory;
+
+    public RedshiftTableStatisticsReader(ConnectionFactory connectionFactory)
+    {
+        this.connectionFactory = connectionFactory;
```
Comment (electrum) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftTableStatisticsReader.java:45]: Check for null

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftTableStatisticsReader.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        private final Handle handle;
+
+        public StatisticsDao(Handle handle)
+        {
+            this.handle = requireNonNull(handle, "handle is null");
+        }
+
+        Optional<Long> getRowCountFromPgClass(String schema, String tableName)
+        {
+            return handle.createQuery("" +
```
Comment (electrum) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftTableStatisticsReader.java:141]: I'd format on one line like the others (length seems about the same)

**Review by electrum — APPROVED** [review]

**Comment by findepi** [comment]
> This was co-authored by: @alexjo2144 @ebyhr @findepi @findinpath @grantatspothero @hashhar @jirassimok @kokosing @losipiuk @Praveen2112 @raunaqmorarka @skrzypo987 @ssheikin @wendigo

Add this to each commit's message.

File: plugin/trino-redshift/pom.xml [hunk 7]
```diff
@@ -23,12 +23,22 @@
             <artifactId>trino-base-jdbc</artifactId>
```
Comment (kokosing) [inline plugin/trino-redshift/pom.xml:23]: This change is backward incompatible change. It is changing semantics of Redshift connector. One could say that after this change it is a new different Redshift connector. Most of the existing production usages of Redshift will now require a migration to make Trino upgrade. Such migration is a binary step. Either user is adjusted to use 403 Redshift connector or 404. There is no middle ground, so rolling back such migration is also difficult.

**Comment by martint** [comment]
> Add this to each commit's message.

Done

File: plugin/trino-redshift/pom.xml (hunk 7 above)
Comment (hashhar) [reply plugin/trino-redshift/pom.xml:23]: Let's restore old one as `redshift-legacy`?

Comment (kokosing) [reply plugin/trino-redshift/pom.xml:23]: I am not sure if it is enough. I would prefer to rename the existing one to mark it is deprecated (legacy) and create a new one with unique name (`redshift-v2`?). Then after some time once legacy is removed I would rename the one to `redshift`)
